### ISSUE TITLE
Switch to run-time index space types

### ISSUE THREAD

**Issue by Krzmbrzl** [ValeevGroup/SeQuant#123]
Right now, the different index space types are hard-coded in the source code and those definitions seem to try to cover all (reasonable) use-cases. However, what it all boils down to (it seems) is that the program is able to determine whether or not two given index spaces overlap and if they do, what the overlapping part is.
This is currently done quite conveniently by using bit-masks, but I don't think that it is necessary to define these bitmasks at compile-time.

Instead, I propose letting users define their own index spaces at runtime, without having to use the pre-defined ones which probably contain many spaces that the user isn't actually interested in, but they still have to wrap their head around them to figure out what they are for and whether or not they might be used internally.

If users could simply define their own index spaces and then be sure that the program never uses any other spaces, this would make adoption of SeQuant easier, simply because users can define the environment in the way that they are used to work with and be done with it.

And instead of defining certain checks based on a space's pre-defined type, one could instead work with properties that could be stored as member variables for the individual spaces (or stored somewhere when registering a given space).
For instance, to query whether or not a given index space is occupied in the chosen reference, one could add 
```cpp
enum class Occupation {
	Doubly,
	Never,
	Partially,
};
```
and add a `get_reference_occupation` function to the `Space` class.

This would also alleviate the need to pre-define index spaces for different contexts all at once as this seems to make things a bit more complicated than having dedicated index spaces per application.

**Comment by Krzmbrzl** [comment]
> Although a runtime solution is probably optimal, the static IndexSpace::attr structures are used in almost every piece of code

Which I would argue is part of the root problem here that should be fixed. Having all kinds of codes depend on these pre-defined index spaces explicitly reduces the overall generalizability of the affected routines. For your suggestion of having orthogonal MR index spaces, this would more or less imply having to add an orthogonal code-path to all those routines to also work in the MR case (where appropriate at least).

> Also, having standard single reference spaces predefined probably makes sequant more approachable

You could still have that. Analogously to the approach to the used `convention`, SeQuant could offer a simple method like `define_default_index_spaces` or something along those lines. If you additionally want a type-safe way of dealing with these default index spaces, add an `enum class DefaultIndexSpace` that users can make use of.
However, if they choose not to use the standard spaces, they could. In particular, this would allow them to define exactly those index spaces that they know and that they are used to working with and arguing about. _This_,  I would argue, would make SeQuant more approachable.

> Therefore, I am proposing that multireference spaces should be defined orthogonal to single reference spaces and given prefix labels (MR)which identify them as such. This way, all set operations behaving within a given context will behave reasonably

This would probably improve the situation to some degree. However, see above for my concerns about generalizability of functions that need to work with MR _and_ SR equations.
Furthermore, I am not entirely sure that this would solve all issues with regards to logical set operations on index spaces. SeQuant caters to some more special cases (afaik) like e.g. using cluster-specific virtuals and I am not entirely sure that all these special cases mix well with "non-special" cases where these kinds of things don't exist.
Just think of a user using an index of the SR `all` space and then after processing is expecting to get either `occupied` or `unoccupied` index spaces, but ends up seeing a space related to cluster-specific virtuals because somewhere internally during determining an overlap of index spaces that's what popped up.
Now, this is a made-up example and I don't have a clue whether the current implementation could yield a result like this. But I think it serves as a highlight of what can go wrong with having hard-coded index spaces for all possible applications hard-code in a single context (SR) already.

Finally, from a practical point of view, I am not yet entirely convinced that switching to runtime index spaces would actually be that much more work than introducing an orthogonal set of index spaces :thinking:

**Comment by evaleev** [comment]
some notes to clarify why the particular design was chosen ... this hopefully can help to decide how to improve it

Truths:
- SeQuant must support working with multiple vacua within same application. Sometimes it is convenient to convert from one vacuum to another.
- SeQuant should be usable without defining/"loading" anything. This is the default (physical vacuum, no spaces pre-registered). Most examples in README.md assume "bare" SeQuant state.
- For Fermi and other vacua the "base" space is partitioned into primal subspaces (occ, uocc). Depending on the application, one may need more than 2 primal subspaces. But even with 2 subspaces one must deal with spaces that are unions of two or more primal subspaces. Thus one must be able to compute intersection and union of 2 spaces.
- Spaces must be further combined with quantum number-like attributes (spin projection quantum number `m_s`, particle type for polaritonics/NEO methods, etc.). Hence `IndexSpace` is a combination of `IndexSpace::Type` and `IndexSpace::Attr`. The quantum numbers compose with set-theoretic operations trivially.
- Wick theorem and other symbolic transformations require being able to compute overlap of 2 spaces (this is implemented using space intersect operation) AND generate a new (dummy) index supported by a given index space. To make expressions including such dummy indices generate human-readable representation one must keep a registry of index labels for each space ... otherwise we can use same base index for any dummy index, but then it will be impossible to interpret e.g. latex representation because it does not encode the type of an index space, it relies on the fact that the base label has 1-to-1 map to the index space.

Possible misconception:
- One does not need to use the predefined (standard) space types. `IndexType` can be constructed directly. What's hardwired is the representation of `IndexType`, not which `IndexType` SeQuant supports; the reason is that it's not clear how to defined set-theoretic ops on partitioned 1-particle Hilbert space.

Alternatives to current design:
- note that the use of hardwired spaces (e.g. `IndexSpace::occupied`) is not mandatory, `IndexSpace` objects can be created programmatically. For simple expression manipulation existence of hardwired instances is not required. Some pieces of code, however, are either require such conventional spaces (e.g. anything in mbpt ... otherwise how one can define SRCC operators?).
- The only piece of code outside of MBPT that uses standard spaces is the Wick engine. We can work around this by allowing the user to inject their own custom logic (such as `qp{annihilator, creator}_space()`). I am not so sure this will be used by anyone though. The logic is quite tricky to get right. To determine whether a particular pair of `Op`s can be contracted one needs to first find the creator/annihilator subspaces of the corresponding ops (which is clearly vacuum dependent) and then compute their intersect.

@Krzmbrzl can you elaborate on what are the concrete issues that need to be addressed here? E.g. what types of uses we do not support that need to be supported? Could you propose a redesign of `IndexSpace` that will not sacrifice the existing use cases?

P.S. Clearly lack of documentation and/or the need to support advanced use cases are significant issues. I don't know how to make things simpler without sacrificing the ability to handle the advances uses.

@cmasteran1 I'm not sure if separating MR spaces is a good idea. What's the problem with having an active space? Simply because we have more possible combinations of primal spaces?

**Comment by Krzmbrzl** [comment]
> SeQuant must support working with multiple vacua within same application. Sometimes it is convenient to convert from one vacuum to another.

Does that mean defining a set of equations with regards to e.g. Fermi vacuum and then while manipulating those expressions it can be beneficial to suddenly switch to e.g. physical or product vacuum? Could you sketch a rough example in which case this would be done? I'm having a bit of a hard time imagining where this would be useful/possible and how this would be done in the current implementation :thinking: 

> SeQuant should be usable without defining/"loading" anything. This is the default (physical vacuum, no spaces pre-registered). Most examples in README.md assume "bare" SeQuant state.

Technically (at least afaict), SeQuant currently doesn't work that way. One has to call the routine that sets up a default convention, which in turn will e.g. set up default labels for all kinds of different index spaces. In the same spirit, I would imagine having a function (or a handful of functions, depending on the use case) that would pre-configure a common set of index spaces that users will typically use.

> For Fermi and other vacua the "base" space is partitioned into primal subspaces (occ, uocc). Depending on the application, one may need more than 2 primal subspaces. But even with 2 subspaces one must deal with spaces that are unions of two or more primal subspaces. Thus one must be able to compute intersection and union of 2 spaces.

As long as we keep the idea of using bitsets as a property of an index space, computing unions and intersections should be essentially the same as it currently is.

> Spaces must be further combined with quantum number-like attributes (spin projection quantum number m_s, particle type for polaritonics/NEO methods, etc.). Hence IndexSpace is a combination of IndexSpace::Type and IndexSpace::Attr. The quantum numbers compose with set-theoretic operations trivially.

I think I have thus far only seen this being used for the spin projection quantum number $m_s$ in order to explicitly assign a given index (space) to being either alpha or beta spin.
In those cases it seemed to me that the quantum number property of index spaces is a somewhat orthogonal concept to index spaces themselves. What I mean with that is that it would be possible to first define index spaces without any quantum numbers and then derive spaces from those with all possible quantum number associations. E.g. first define a space for occupied indices and from that then derive spaces for occupied+alpha and occupied+beta.

Does this observation generalize to other use cases?

---

> One does not need to use the predefined (standard) space types. IndexType can be constructed directly. What's hardwired is the representation of IndexType, not which IndexType SeQuant supports; the reason is that it's not clear how to defined set-theoretic ops on partitioned 1-particle Hilbert space.

So what you are saying with that is that the only thing that is fixed is that I have to represent my index spaces by means of a bitset in order to be able to have a way of computing index space intersections / unions?
That would essentially already be what I was requesting :eyes: 

---

> Some pieces of code, however, are either require such conventional spaces (e.g. anything in mbpt ... otherwise how one can define SRCC operators?).

That's the culprit. So far it seems to me that the use of SeQuant rather heavily relies on users making use of the `mbpt` domain code as it implements things like spin tracing, which usually is quite important in applications.
Maybe my view is a bit biased here as I myself am coming from the Coupled Cluster corner, but at least for me it seems like use of the `mbpt` code is mandatory. And thus not using the pre-defined index spaces seems tricky.

---

### My proposition

Originally, I indeed thought the use of the pre-defined was mandatory. Now that I know that it is now (at least formally), I think this issue is more about reducing those parts of the code that currently explicitly require use of those hard-coded spaces.

You mentioned the Wick theorem as being one of those places. Specifically, I think you are referring to
https://github.com/ValeevGroup/SeQuant/blob/7f4b568f601034cc4c615965f80ff68800164e59/SeQuant/core/op.hpp#L205-L223
and the other qp-related functions. Taking this particular function as an example, it seems what we really need to know is the bidirectional association space <-> occupancy in vacuum.

I think this could be solved without resorting to the use of hard-coded index spaces (space types) by having a central entity (e.g.  as a singleton) that I would call an `IndexSpaceManager` that could look something like this:
```cpp
enum class ReferenceOccupation {
	Always,
	Partially,
	Never,
};

class IndexSpaceManager {
public:
	IndexSpace for_occupation(ReferenceOccupation occ, Vacuum vac) const;

    void register(IndexSpace space);
};
```
this would require used index spaces to be registered to the `IndexSpaceManager` instance (could be automated by doing this in the ctor if the `IndexSpace` class) and then you can query the manager for an index space that is the union of all registered index spaces that have the given occupancy for the provided vacuum.

This involves some subtleties with regards to whether there should be an order-requirement for the registration of index spaces (first the union of all spaces with a given occupancy and then the individual spaces) or not (in which case the manager would have to be able to automatically create a union space as soon as it sees a second space with the same occupancy - which should of course be replaced if later on a union space is provided explicitly).

The information about which space has what occupation in the given Vacuum should come from the index spaces themselves. Thus, I would propose a query function like
```
ReferenceOccupation IndexSpace::occupation(Vacuum vac) const;
```
For simple cases the entire processing involving a given index space will only use a single vacuum (I think?) and thus this can return a constant occupation ignoring the `vac` parameter (maybe `assert`ing that it is the expected vacuum).
For more complex scenarios these functions would have to do some space-specific processing. I would propose that these more complex scenarios are either handled by creating subclasses of `IndexSpace`  or by making this query function a free function templated on some index space tag that can be specialized a needed.
(The mapping to physical vacuum is always trivial though as it simply is `ReferenceOccupation::Never`).

The same concept seems also applicable to operator creation in `mbpt`, e.g. in
https://github.com/ValeevGroup/SeQuant/blob/7f4b568f601034cc4c615965f80ff68800164e59/SeQuant/domain/mbpt/mr.cpp#L133-L139
where (always w.r.t. product vacuum) `unocc` is simple the union of all partially occupied spaces and all those which are never occupied in the reference. And `occ` is likewise the union of all partially occupied with all always occupied indices.

In more specialized cases, e.g. for the construction of the geminal kernel as in
https://github.com/ValeevGroup/SeQuant/blob/7f4b568f601034cc4c615965f80ff68800164e59/SeQuant/domain/mbpt/mr/op.impl.cpp#L57-L68
I think the user should simply provide the necessary space to use instead of the currently hard-coded `other_unoccupied`. Should something like this (or analogously for `frozen_occupied`) be required more often or somewhere nested more deeply, one would have to think of a second new property of the `IndexSpace` class that would allow code to query for these special spaces explicitly from `IndexSpaceManager`.


#### Why bother?

I think that with my suggestion from above (or a suitable generalization of it as I have most likely missed cases that are currently not covered by it) one could get rid of many, if not most or even all uses of hard-coded index space types in SeQuant's code base.

However, introducing such a framework adds some complexity into SeQuant's core implementation and also a (likely (very) small) runtime penalty as we now have to perform lookups at runtime instead of letting the compiler know at compile time what it is going to see. So why bother with all this?

The main benefits that I see in removing dependence of hard-coded index space types are
- Every application (as in applying SeQuant's capabilities to derive a given method) will only ever have to worry about the specific index spaces that it is actually going to use. Currently, choosing which of the existing pre-defined spaces is the one who will make SeQuant do what you want it to do can be a bit overwhelming (and always using the pre-defined operators from `mbpt` is likely not always an option)
- Set operations on index spaces behave as expected. This is the problem @cmasteran1 seems to have encountered. Due to the pre-defined space types trying to be very general, in certain cases the set operations don't yield the expected outcome (from the user's POV). The provided example was "union of the occupied and unoccupied spaces should give you the all space" in the context of simple SR applications. However, that is currently not the case.
- Makes "reasoning" in SeQuant's core implementation more explicit. With that I mean it will be forced to more closely query for the actual property it is looking for instead of using a hard-coded index space type that is know to (implicitly) have the desired property. Admittedly, this is not _that_ important but it might facilitate new developers to get an understanding of what is actually going on.

**Comment by Krzmbrzl** [comment]
This has been implemented via #187
